airflow/providers/amazon/aws/hooks/s3.py

```diff
@@ -469,7 +469,7 @@ def get_key(self, key: str, bucket_name: Optional[str] = None) -> S3Transfer:
             verify=self.verify,
         )
         obj = s3_resource.Object(bucket_name, key)
-        obj.load()
+        obj.load(**self.extra_args)
```

Comment:
Are you sure that you could provide anything to [S3.Object.load()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Object.load) method?

Also please have a look [this](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#s3-transfers). `extra_args` expected only use in this s3 client methods

- [S3.Client.download_file()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.download_file)
- [S3.Client.download_fileobj()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.download_fileobj)
- [S3.Client.upload_file()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.upload_file)
- [S3.Client.upload_fileobj()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.upload_fileobj)

And I do not sure `extra_args` values fully compatible with `obj.load()`.

Reply:
I would suggest to remove this part
https://github.com/apache/airflow/blob/4a5e247536b2b7cb304f2f8dc5961d94278affe1/airflow/providers/amazon/aws/hooks/s3.py#L869-L877

At that moment in this method first we try to check if object exists before download and raise an error if not exists.
This check could cause an issues in some cases.

Might be better do not use `boto3.resource("s3").Object` at all and just try to download object by `boto3.client("s3")` instead if any error happen we still can catch them and inform users.

Example:

```python
        with NamedTemporaryFile(dir=local_path, prefix='airflow_tmp_', delete=False) as local_tmp_file:
            err = None
            try:
                self.get_conn().download_fileobj(
                    bucket_name,
                    key,
                    local_tmp_file,
                    ExtraArgs=self.extra_args,
                    Config=self.transfer_config,
                )
            except Exception as e:
                err = True
                if isinstance(e, ClientError):
                    ...
                    # Do some error handling here
                raise
            finally:
                if err:
                    if not local_tmp_file.closed:
                        local_tmp_file.close()
                    Path(local_tmp_file.name).unlink()
```

WDYT @moreaupascal56 

cc: @o-nikolas @vincbeck @ferruzzi

Reply:
A few things are getting conflated in this thread.
#### 1) The PR code change:
It looks like you're just unpacking the contents of `extra_args` into the `load` call. I agree with @Taragolis that I'm not sure `obj.load` expects all the things it could receive from that dict (or if any of them would even help with encryption). Object.load ([docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Object.load)) basically just calls s3client `head_object` ([docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object)) which is one of those weird S3 APIs:
>A HEAD request has the same options as a GET action on an object. The response is identical to the GET response except that there is no response body. Because of this, if the HEAD request generates an error, it returns a generic 404 Not Found or 403 Forbidden code. It is not possible to retrieve the exact exception beyond these error codes.

To handle encryption when using that API you have to set request headers (see again the previous doc link):

>If you encrypt an object by using server-side encryption with customer-provided encryption keys (SSE-C) when you store the object in Amazon S3, then when you retrieve the metadata from the object, you must use the following headers:

    x-amz-server-side-encryption-customer-algorithm
    x-amz-server-side-encryption-customer-key
    x-amz-server-side-encryption-customer-key-MD5

#### 2) Andrey's suggested change:

> At that moment in this method first we try to check if object exists before download and raise an error if not exists.
This check could cause an issues in some cases.

Just so that I fully understand your proposal, what are some cases you're thinking of? Also this change seems out of scope for what @moreaupascal56 is trying to achieve?

Also, also, for your suggested code snippet, you're explicitly closing and unlinking the temp file, do you not just the context manager to do so?

Reply:
> Just so that I fully understand your proposal, what are some cases you're thinking of?

Both general case and @moreaupascal56 case. My idea is just use more common pythonic approach: **Easier to Ask for Forgiveness than Permission**. 

Most of the code was add about 2 years ago and most probably it just migrated from < Airflow 2 and I thought current implementation do not cover all use cases.

Initially by the code method tried get information about object but do not pass any of the `boto3.s3.transfer` Download Arguments there:  `'ChecksumMode', 'VersionId', 'SSECustomerAlgorithm', 'SSECustomerKey', 'SSECustomerKeyMD5', 'RequestPayer', 'ExpectedBucketOwner'`.

If user want specific _VersionId_ it actually check latest _VersionId_ or delete marker (need to check just my thoughts).

Example
```
hook = S3Hook(aws_conn_id="aws_default", extra_args={"VersionId": "AAAAAAAA"})
hook.download_file(key="some/awesome/key.parquet", bucket_name="awesome-bucket")
```

So what might happen here?
`s3_obj = self.get_key(key, bucket_name)` might return object with Version ID = "AAAAAAAA" or "BBBBBBBB"

After that we tried to download file and pass ExtraArgs = `{"VersionId": "AAAAAAAA"}` and I thought it is possible that get some funny error if `s3_obj` referenced to Version ID = "BBBBBBBB".

So my idea get rid of call `self.get_key` and just use [S3.Client.download_fileobj()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.download_fileobj) instead which might handle passing correct headers for customer-provided encryption and/or Version ID

> you're explicitly closing and unlinking the temp file, do you not just the context manager to do so?

As I understand current code correct `NamedTemporaryFile` use for create unique file inside Airflow Worker, and explicit set that this file shouldn't be deleted because this method returns file path to downloaded file.

In my snippet I just add manually close file if any error happen because we still in context manager and unlink them because if some error happen this file not useful anymore and we could remove it.
Some sample of error which could keep some unwanted files, all of them not related to permission:
- Connection error in the middle of download file
- User tried download 20 GiB file inside Fargate temp directory and get "No space left on device"

If everything alright than context manager close the file

Reply:
The context manager will clean up the file even in the case of exception (see [here](https://stackoverflow.com/a/28158006/1055702)) so there is no need to do any file handling in a finally. This is the beauty of Context Managers! :smiley: